core/templates/services/exploration-improvements-task-registrar.service.ts

```diff
@@ -0,0 +1,424 @@
+  private expVersion: number;
+  private expStats: ExplorationStats;
+
+  private taskRegistry: Map<string, TaskRegistry> = new Map();
+  private statsRegistry: Map<string, StatsRegistry> = new Map();
+
+  private tasksByType: Map<ExplorationTaskType, ExplorationTask[]> = (
+    new Map(ImprovementsConstants.TASK_TYPES.map(t => [t, []])));
+
+  private initializationChangesAreFlushed: boolean = false;
+
+  constructor(
+      private explorationTaskObjectFactory: ExplorationTaskObjectFactory) {}
+
+  initialize(
+      expId: string,
+      expVersion: number,
+      states: States,
+      expStats: ExplorationStats,
+      openTasks: ExplorationTask[],
+      resolvedTaskTypesByStateName: Map<string, ExplorationTaskType[]>,
+      stateAnswerStats: Map<string, AnswerStats[]>,
+      cstPlaythroughsByStateName: Map<string, CstPlaythrough[]>,
+      eqPlaythroughsByStateName: Map<string, EqPlaythrough[]>,
+      misPlaythroughsByStateName: Map<string, MisPlaythrough[]>) {
+    this.validateInitializeArgs(
+      expId, expVersion, states, expStats, openTasks,
+      resolvedTaskTypesByStateName, stateAnswerStats,
+      cstPlaythroughsByStateName, eqPlaythroughsByStateName,
+      misPlaythroughsByStateName);
+
+    this.expId = expId;
+    this.expVersion = expVersion;
+    this.expStats = expStats;
+
+    const allStateNames = states.getStateNames();
+    this.initializeStatsRegistry(
+      allStateNames, stateAnswerStats, cstPlaythroughsByStateName,
+      eqPlaythroughsByStateName, misPlaythroughsByStateName);
+    this.initializeTaskRegistry(
+      allStateNames, openTasks, resolvedTaskTypesByStateName);
+
+    for (const [stateName, taskRegistry] of this.taskRegistry.entries()) {
+      const statsRegistry = this.statsRegistry.get(stateName);
+      taskRegistry.forEach(t => this.refreshTask(t, statsRegistry));
+    }
+  }
+
+  flushInitializationChanges(): ITaskEntryPayloadDict[] {
+    if (this.initializationChangesAreFlushed) {
+      return [];
+    }
+    this.initializationChangesAreFlushed = true;
+
+    const payload = [];
+    for (const taskRegistry of this.taskRegistry.values()) {
+      if (taskRegistry.didHbrTaskStatusChange()) {
+        payload.push(taskRegistry.hbrTask.toPayloadDict());
+      }
+    }
+    return payload;
+  }
+
+  onStateAdd(stateName: string): void {
+    const newTasks = ImprovementsConstants.TASK_TYPES.map(
+      taskType => this.explorationTaskObjectFactory.createNewObsoleteTask(
+        this.expId, this.expVersion, taskType, stateName));
+    this.taskRegistry.set(stateName, new TaskRegistry(newTasks));
+    this.statsRegistry.set(stateName, new StatsRegistry());
+    newTasks.forEach(t => this.tasksByType.get(t.taskType).push(t));
+  }
+
+  onStateDelete(stateName: string): void {
+    const taskRegistry = this.taskRegistry.get(stateName);
+    taskRegistry.forEach(t => t.markAsObsolete());
+
+    // eslint-disable-next-line dot-notation
+    this.taskRegistry.delete(stateName);
```

Comment:
You could perhaps use a name that's not "delete", if that's what's causing problems (I think it's a reserved JS operator).

If that's not the case -- what's the "dot-notation issue" here? I don't really see any use of [brackets](https://eslint.org/docs/rules/dot-notation) to trigger it.

Reply:
Yeah this seems to be an unfortunate design choice in JavaScript maps, which use `delete` as a method even though it is a reserved keyword ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete)).

This lint was the easiest solution I could find to bypass the error (which itself isn't even correct..?)

Reply:
That is weird. Maybe add a detailed comment explaining what is going on with that error message, and file a bug on https://github.com/eslint/eslint/issues ?